Switch unix port to common select module, and add POSIX optimisations to that module

dpgeorge opened the pull request:
The unix port has a custom `select` module which only works with objects that have a file descriptor, eg files and sockets.  On the other hand, bare metal ports use the common `extmod/modselect.c` implementation of the select module that supports polling of arbitrary objects, as long as those objects provide a `MP_STREAM_POLL` in their ioctl implementation (which can be done in Python).

This PR removes the unix-specific select module implementation and makes unix use the common one.  And then it also adds an optimisation so that the system `poll` function is used when polling objects that do have a file descriptor.  If you register both file-descriptor-based objects, and non-file-descriptor-based objects with `select.poll()` then the following occurs:
- the system poll is called for all file-descriptor-based objects with a timeout of 1ms
- then the bare-metal polling implementation is used for remaining objects, which calls into their ioctl method (which can be in C or Python)

In the case where all objects have file descriptors, the system poll is called with the full timeout requested by the caller.  That makes it as efficient as possible in the case everything has a file descriptor.

Benefits of this approach:
- all ports use the same `select` module implementation
- the unix port now supports polling of all objects and matches bare metal implementations
- still efficient for existing cases where only files and sockets are polled (on unix)
- bare metal implementation does not change
- polling of SSL objects will now work on unix by calling in to the ioctl method on SSL objects; this is required for #11897

An additional follow-up optimisation would be to add a special case to handle polling of SSL sockets so it can poll the underlying file descriptor of the socket.  This requires some tricks to swap read/write polling depending on the state of the SSL transport.

dpgeorge committed:
unix/modselect: Remove unix-specific implementation of select module.

The unix port has a custom select module which only works with objects that
have a file descriptor, eg files and sockets.  On the other hand, bare
metal ports use the common extmod/modselect.c implementation of the select
module that supports polling of arbitrary objects, as long as those objects
provide a MP_STREAM_POLL in their ioctl implementation (which can be done
in C or Python).

This commit removes the unix-specific code and makes unix use the common
one provided by extmod/modselect.c instead.  All objects with file
descriptors implement MP_STREAM_POLL so they continue to work.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
unix/modsocket: Add poll support for missing ERR,HUP,NVAL poll values.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
extmod/vfs_posix_file: Add poll support for missing ERR,HUP,NVAL values.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
extmod/modselect: Abstract out a poll_set_t struct and functions.

To make it easier to extend and modify this polling implementation.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
extmod/modselect: Factor low-level polling code into common function.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
extmod/modselect: Add optimisation to use system poll when possible.

A previous commit removed the unix-specific select module implementation
and made unix use the common one.

This commit adds an optimisation so that the system poll function is used
when polling objects that have a file descriptor.  With this optimisation
enabled, if code registers both file-descriptor-based objects, and non-
file-descriptor-based objects with select.poll() then the following occurs:

- the system poll is called for all file-descriptor-based objects with a
  timeout of 1ms

- then the bare-metal polling implementation is used for remaining objects,
  which calls into their ioctl method (which can be in C or Python)

In the case where all objects have file descriptors, the system poll is
called with the full timeout requested by the caller.  That makes it as
efficient as possible in the case everything has a file descriptor.

Benefits of this approach:

- all ports use the same select module implementation

- the unix port now supports polling of all objects and matches bare metal
  implementations

- it's still efficient for existing cases where only files and sockets are
  polled (on unix)

- the bare metal implementation does not change

- polling of SSL objects will now work on unix by calling in to the ioctl
  method on SSL objects (this is required for asyncio ssl support)

Note that extmod/vfs_posix_file.c has poll disable when the optimisation is
enabled, because the code is not reachable when the optimisation is used.

Signed-off-by: Damien George <damien@micropython.org>

jimmo commented on extmod/modselect.c:
```diff
@@ -26,62 +26,223 @@
     }
 }
 
 // poll each object in the map
```
Unless you know (from knowing about this PR) what's going on here, the concept of a "single round" might not be that obvious. The commit description doesn't explain it either.

I'd probably suggest renaming `poll_group_poll_all` to `poll_group_poll_until_ready_or_timeout` and then perhaps this function should be `poll_group_poll_once`.

dpgeorge replied:
Renamed as suggested.

jimmo commented on extmod/modselect.c:
```diff
@@ -26,62 +26,223 @@
     mp_obj_t obj;
     mp_uint_t (*ioctl)(mp_obj_t obj, mp_uint_t request, uintptr_t arg, int *errcode);
+    #if MICROPY_PY_SELECT_POSIX_OPTIMISATIONS
+    struct pollfd *pollfd;
```
This needs a comment to explain that pollfd is mutually exclusive with flags and flags_ret. (I think that's what you're getting at with the `nonsys_` prefix, but it would be clearer to just spell it out and remove the prefix... although I guess the prefix stops you using the non-posix-opt code accidentally?).

jimmo replied:
Also I think it would be good to be explicit that the pollfd member here points to the element in `poll_group_t::pollfds`.

dpgeorge replied:
Comments added, and `nonsys_` prefix changed to `nonfd_`.

Also, `flags` renamed to `events` and `flags_ret` renamed to `revents` to make it more consistent and clearer (a `flags` argument is used elsewhere for completely unrelated things).

jimmo commented on extmod/modselect.c:
```diff
@@ -107,6 +268,69 @@ STATIC mp_uint_t poll_map_poll(mp_map_t *poll_map, size_t *rwx_num) {
+        MP_THREAD_GIL_EXIT();
+
+        // Compute the timeout.
+        int t = 1;
```
This `1` is potentially an important tuning parameter and should be a defined constant.

dpgeorge replied:
This has been made a macro (private to this file, but in the future could be made public and configurable).

jimmo commented on extmod/modselect.c:
```diff
@@ -107,6 +268,69 @@ STATIC mp_uint_t poll_map_poll(mp_map_t *poll_map, size_t *rwx_num) {
+        // Compute the timeout.
+        int t = 1;
+        if (poll_group_all_are_fds(poll_group)) {
+            if (timeout == (mp_uint_t)-1) {
```
`// All our pollables are file descriptors, so we can use blocking `poll` and let it handle the timeout.`

dpgeorge replied:
Comment added.

andrewleech commented:
I don't have anything particularly constructive to add from a review other than it looks really good thanks! This will be great for the unix port going forward.

Is this something that would also help with integrating https://github.com/micropython/micropython/pull/6125 into all ports in the future? Or does an event system sit alongside select?

jimmo commented on extmod/modselect.c:
```diff
@@ -107,6 +268,69 @@ STATIC mp_uint_t poll_map_poll(mp_map_t *poll_map, size_t *rwx_num) {
+            return n_ready;
+        }
+
+        mp_handle_pending(true);
```
I was going to suggest that we should also do an explicit sleep here, and also why not use `MICROPY_EVENT_POLL_HOOK`. But that's actually what `MICROPY_EVENT_POLL_HOOK` does, so perhaps we should use it.

Perhaps a smarter option would be to only do the sleep if none of the pollables are file descriptors (i.e. we didn't get the sleep from `poll`).

dpgeorge replied:
Comment added telling why `MICROPY_EVENT_POLL_HOOK` is not used.

dpgeorge commented:
There are some very minor differences than the unix port will see with this PR:
- `poll.register()` now returns `None` (which is CPython compatible) instead of True/False (which was a MicroPython extension)
- `select.poll()` no longer accepts an optional alloc argument (again, this was a MicroPython extension, mainly for old uasyncio)
- `poll.poll()` now returns `[]` instead of `()` when there are no objects ready (on should prefer using `ipoll()` anyway for efficiency)

jimmo commented on extmod/modselect.c:
```diff
@@ -26,62 +26,223 @@
 } poll_obj_t;
 
-STATIC void poll_map_add(mp_map_t *poll_map, const mp_obj_t *obj, mp_uint_t obj_len, mp_uint_t flags, bool or_flags) {
+// A group of pollable objects.
```
Initially I was confused by what a "poll group" was, I think I expected maybe you'd group-by the poll-style (i.e. fd or no-fd). But there's never more than one group, rather it's a convenient way of bundling the map with the list of pollfds (and metadata for that list).

(I think "list" or "set" is a better term than "group", because "group" implies they're grouped by something).

dpgeorge replied:
Renamed `poll_group_t` to `poll_set_t`.

jimmo reviewed (commented):
Thanks, this is a great improvement! Very helpful for running `aioble` on unix.

jimmo commented on extmod/modselect.c:
```diff
@@ -107,6 +268,69 @@ STATIC mp_uint_t poll_map_poll(mp_map_t *poll_map, size_t *rwx_num) {
+
+        if (n_ready == -1) {
+            int err = errno;
+            if (err != EINTR) {
```
should this also be `EAGAIN` ?

Also, the old code used `MP_HAL_RETRY_SYSCALL` ?

dpgeorge replied:
Comment added about PEP 475 and `MP_HAL_RETRY_SYSCALL`.

dpgeorge committed:
tests/extmod: Add coverage tests for select module.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
extmod/modselect: Remove undocumented support for flags arg to poll.

The signature of this method was poller.poll(timeout=-1, flags=0, /) but
the flags argument was not documented and is not CPython compatible.  So
it's removed in this commit.

(The optional flags remains for the ipoll() method, which is documented.)

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
Enough tests have now been added so the coverage CI now passes.

jimmo commented:
Looks good! Nice work with the coverage tests.

dpgeorge committed:
tests/extmod: Skip select/socket tests if they can't create UDP socket.

Some targets (eg PYBV10) have the socket module but are unable to create
UDP sockets without a registered NIC.  So skip UDP tests on these targets.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
New tests have been tested on bare-metal (stm32) and adjusted so they skip or pass there as well.

All comments above are now addressed, it was purely renaming of variables/functions and adding comments.
